### PR TITLE
consensus/clique: add some missing checks (#22836)

### PR DESCRIPTION
**Pull request by wanwiset25**
# Proposed changes
Add checks to make sure gasLimit is within valid range as advised by the auditing team.
ref: https://github.com/ethereum/go-ethereum/pull/22836

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [x] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [ ] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [x] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

### CODE REVIEW

File: consensus/clique/clique.go [hunk 1]
```diff
@@ -347,6 +357,15 @@ func (c *Clique) verifyCascadingFields(chain consensus.ChainReader, header *type
 	if parent.Time.Uint64()+c.config.Period > header.Time.Uint64() {
 		return ErrInvalidTimestamp
 	}
+	// Verify that the gas limit remains within allowed bounds
+	diff := int64(parent.GasLimit) - int64(header.GasLimit)
+	if diff < 0 {
+		diff *= -1
```
Comment (gzliudan) [inline consensus/clique/clique.go:363]: `diff = -diff` maybe quicker.

Comment (wanwiset25) [reply consensus/clique/clique.go:363]: Maybe you're right, but I think it's very small optimization. I don't want to modify the original author's code. In the future people may be confused why they are different.

**Review by gzliudan — APPROVED** [review]

File: consensus/clique/clique.go [hunk 2]
```diff
@@ -316,6 +317,15 @@ func (c *Clique) verifyHeader(chain consensus.ChainReader, header *types.Header,
 			return errInvalidDifficulty
 		}
 	}
+	// Verify that the gas limit is <= 2^63-1
+	cap := uint64(0x7fffffffffffffff)
+	if header.GasLimit > cap {
+		return fmt.Errorf("invalid gasLimit: have %v, max %v", header.GasLimit, cap)
+	}
```
Comment (gzliudan) [inline consensus/clique/clique.go:324]: Below is the latest code in ethereum:
```go
	// Verify that the gas limit is <= 2^63-1
	if header.GasLimit > params.MaxGasLimit {
		return fmt.Errorf("invalid gasLimit: have %v, max %v", header.GasLimit, params.MaxGasLimit)
	}
```

**Comment by wanwiset25** [comment]
I updated the logic to match the current go-eth code. 

Thanks for the suggestion @gzliudan

**Review by gzliudan — APPROVED** [review]

**Comment by wjrjerome** [comment]
We actually don't need to do any work for the clique consensus. it's not been used anywhere.

**Review by wjrjerome — APPROVED** [review]

**Pull request merged**